### PR TITLE
WIP Modify The Logic Of Replace In The Copy Method

### PR DESCRIPTION
**Pull request by fireworkmarks**

### CODE REVIEW

File: api/object-handlers.go [hunk 1]
```diff
@@ -568,14 +505,25 @@ func (api ObjectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
...
+	} else if r.Header.Get("X-Amz-Metadata-Directive") == "REPLACE" {
+		newMetadata := extractMetadataFromHeader(r.Header)
+		if c, ok := newMetadata["content-type"]; ok {
+			targetObject.ContentType = c
+		} else {
+			targetObject.ContentType = sourceObject.ContentType
+		}
+		targetObject.CustomAttributes = newMetadata
 		targetObject.StorageClass = storageClassFromHeader
 	} else {
```
Comment (cxt90730) [inline api/object-handlers.go]: if Header ‘X-Amz-Metadata-Directive’ not COPY or REPLACE，return error

Comment (cxt90730) [reply api/object-handlers.go]: and the default value should be COPY

File: api/object-handlers.go [hunk 2]
```diff
@@ -620,6 +567,129 @@ func (api ObjectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
...
+	targetBucketName := vars["bucket"]
+	targetObjectName := vars["object"]
+
+	var credential common.Credential
+	var err error
+	if credential, err = checkRequestAuth(api, r, policy.PutObjectAction, targetBucketName, targetObjectName); err != nil {
+		WriteErrorResponse(w, r, err)
+		return
+	}
+	renameSource := r.Header.Get("X-Amz-Copy-Source")
```
Comment (fireworkmarks) [inline api/object-handlers.go]: need change

File: api/object-handlers.go [hunk 3]
```diff
@@ -620,6 +567,129 @@ func (api ObjectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
...
+	//
+	isMultiVersion := api.ObjectAPI.GetObjectMultiVersionInfo(sourceBucketName, sourceObjectName,
+		sourceVersion, credential)
+	if !isMultiVersion {
+		WriteErrorResponse(w, r, ErrInvalidRenameBucketVersion)
+		return
+	}
+
+	// Verify before x-amz-copy-source preconditions before continuing with CopyObject.
+	if err = checkObjectPreconditions(w, r, sourceObject); err != nil {
```
Comment (fireworkmarks) [inline api/object-handlers.go]: this func should do upper

File: api/object-handlers.go [hunk 4]
```diff
@@ -620,6 +567,129 @@ func (api ObjectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
...
+		WriteErrorResponse(w, r, err)
+		return
+	}
+	targetObject := sourceObject
+	
+	//Determine if the renamed object is a folder
+	if string(targetObjectName[len(targetObjectName)-1]) == "/" {
```
Comment (fireworkmarks) [inline api/object-handlers.go]: source should judge too

File: error/api-errors.go [hunk 5]
```diff
@@ -167,6 +171,26 @@ var ErrorCodeResponse = map[ApiErrorCode]ApiErrorStruct{
...
+		HttpStatusCode: http.StatusBadRequest,
+	},
+	ErrInvalidRenameObjectName: {
+		AwsErrorCode:   "InvalidRenameObjectName",
+		Description:    "Object name of rename source cannot be folder.",
+		HttpStatusCode: http.StatusBadRequest,
+	},
+	ErrInvalidRenameBucketVersion: {
+		AwsErrorCode:   "InvalidRenameBucketVersion",
+		Description:    "Bucket of rename source cannot be multiversion.",
```
Comment (fireworkmarks) [inline error/api-errors.go]: need change

File: storage/object.go [hunk 6]
```diff
@@ -325,6 +325,20 @@ func (yig *YigStorage) GetObjectInfo(bucketName string, objectName string,
 	return
 }
 
+// TODO：Determine here whether it is a multi-version control object or bucket
+func (yig *YigStorage) GetObjectMultiVersionInfo(bucketName string, objectName string,
+	version string, credential common.Credential) bool {
+
+	bucket, err := yig.MetaStorage.GetObject(bucketName, objectName, true)
```
Comment (fireworkmarks) [inline storage/object.go]: should be GetObject

File: api/object-handlers.go [hunk 7]
```diff
@@ -620,6 +567,129 @@ func (api ObjectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
...
+
+	sourceObject, err := api.ObjectAPI.GetObjectInfo(sourceBucketName, sourceObjectName,
+		sourceVersion, credential)
+	if err != nil {
+		WriteErrorResponseWithResource(w, r, err, renameSource)
+		return
+	}
+
+	//
+	isMultiVersion := api.ObjectAPI.GetObjectMultiVersionInfo(sourceBucketName, sourceObjectName,
```
Comment (fireworkmarks) [inline api/object-handlers.go]: should check bucket and return entity

**Pull request closed without merging**